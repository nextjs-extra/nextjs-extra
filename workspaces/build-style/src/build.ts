import { readFile } from "fs/promises";
import * as glob from "glob";
import { resolve } from "path";
import { promisify } from "util";
import { outputFile } from "fs-extra";

const globSync = promisify(glob);

const getNameRegex = /(?<name>[^/]+)\.component\.s?css$/;

export async function build({
  globPattern,
  cwd,
  outStyle,
  outVars,
  variables,
}) {
  const files = await globSync(globPattern, {
    cwd,
  });

  let scss = "";
  const vars = new Set<string>();

  for (const file of files) {
    const content = await readFile(resolve(cwd, file), "utf-8");

    const name = getNameRegex.exec(file)?.groups?.name;

    scss += `/* start file: ${file} */\n`;

    if (name) {
      scss += `${name} {\n`;
    }
    scss += content;
    if (name) {
      scss += `}\n`;
    }
    scss += `/* end file: ${file} */\n\n`;

    const matches = content.matchAll(/var\(--(?<var>[^,)]+)/g);

    for (const match of matches) {
      if (match.groups?.var) {
        vars.add(match.groups.var);
      }
    }
  }

  let varsFile = ":root {\n";
  const varsArray = Array.from(vars);
  varsArray.sort();

  const variablesCopy = { ...variables };

  for (const varName of varsArray) {
    if (variables[varName]) {
      varsFile += `  --${varName}: ${variables[varName]};\n`;
      delete variablesCopy[varName];
    } else {
      varsFile += `  /* --${varName}: ;  /* undefined */\n`;
    }
  }
  for (const varName in variablesCopy) {
    varsFile += `  --${varName}: ${variables[varName]}; /* unused */\n`;
  }
  varsFile += "}";

  await outputFile(resolve(cwd, outVars), varsFile);
  await outputFile(resolve(cwd, outStyle), scss);
}
