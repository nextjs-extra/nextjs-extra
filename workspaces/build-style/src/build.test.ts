import { build } from "./build";
import { readFile } from "fs/promises";
import { resolve } from "path";
import { remove } from "fs-extra";

const cwd = resolve(__dirname, "../test");
const buildStyleTestPath = resolve(cwd, "build-style");

describe("build", () => {
  beforeEach(async () => {
    await remove(buildStyleTestPath);
  });

  it("builds", async () => {
    await build({
      globPattern: "**/*.{global,component}.{css,scss}",
      cwd,
      outStyle: "build-style/style.scss",
      outVars: "build-style/vars.scss",
      variables: {
        "color-background": "#ff0000",
        "secondary-color": "#00ff00",
      },
    });

    expect(
      await readFile(resolve(cwd, "build-style/style.scss"), "utf-8")
    ).toMatchSnapshot();
    expect(
      await readFile(resolve(cwd, "build-style/vars.scss"), "utf-8")
    ).toMatchSnapshot();
  });

  afterAll(async () => {
    await remove(buildStyleTestPath);
  });
});
