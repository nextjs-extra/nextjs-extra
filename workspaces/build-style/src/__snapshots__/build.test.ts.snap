// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build builds 1`] = `
"/* start file: main.global.css */
body {
  background: var(--color-background, #fff);
  color: var(--color-text, #000);
  margin-top: auto;
}
/* end file: main.global.css */

/* start file: some-component.component.scss */
some-component {
h1 {
  color: green;
}
}
/* end file: some-component.component.scss */

"
`;

exports[`build builds 2`] = `
":root {
  --color-background: #ff0000;
  /* --color-text: ;  /* undefined */
  --secondary-color: #00ff00; /* unused */
}"
`;
